account-compression/programs/spl-compression/src/lib.rs

```diff
@@ -0,0 +1,550 @@
+    }
+
+    /// Note:
+    /// Supporting this instruction open a security vulnerability for indexers.
+    /// This instruction has been deemed unusable for publicly indexed compressed NFTs.
+    /// Indexing batched data in this way requires indexers to read in the `uri`s onto physical storage
+    /// and then into their database. This opens up a DOS attack vector, whereby this instruction is
+    /// repeatedly invoked, causing indexers to fail.
+    pub fn init_gummyroll_with_root(
+        ctx: Context<Initialize>,
+        max_depth: u32,
+        max_buffer_size: u32,
+        root: [u8; 32],
+        leaf: [u8; 32],
+        index: u32,
+        _changelog_db_uri: String,
+        _metadata_db_uri: String,
+    ) -> Result<()> {
+        let mut merkle_roll_bytes = ctx.accounts.merkle_roll.try_borrow_mut_data()?;
+
+        let (mut header_bytes, rest) =
+            merkle_roll_bytes.split_at_mut(size_of::<ConcurrentMerkleTreeHeader>());
+
+        let mut header = Box::new(ConcurrentMerkleTreeHeader::try_from_slice(&header_bytes)?);
+        header.initialize(
+            max_depth,
+            max_buffer_size,
+            &ctx.accounts.authority.key(),
+            Clock::get()?.slot,
+        );
+        header.serialize(&mut header_bytes)?;
+        let merkle_roll_size = merkle_roll_get_size!(header)?;
+        let (roll_bytes, canopy_bytes) = rest.split_at_mut(merkle_roll_size);
+
+        // Get rightmost proof from accounts
+        let mut proof = vec![];
+        for node in ctx.remaining_accounts.iter() {
+            proof.push(node.key().to_bytes());
+        }
+        fill_in_proof_from_canopy(canopy_bytes, header.max_depth, index, &mut proof)?;
+        assert_eq!(proof.len(), max_depth as usize);
+
+        let id = ctx.accounts.merkle_roll.key();
+        // A call is made to ConcurrentMerkleTree::initialize_with_root(root, leaf, proof, index)
+        let change_log = merkle_roll_apply_fn!(
+            header,
+            id,
+            roll_bytes,
+            initialize_with_root,
+            root,
+            leaf,
+            &proof,
+            index
+        )?;
+        wrap_event(change_log.try_to_vec()?, &ctx.accounts.candy_wrapper)?;
+        emit!(*change_log);
+        update_canopy(canopy_bytes, header.max_depth, Some(change_log))
+    }
+
+    /// Executes an instruction that overwrites a leaf node.
+    /// Composing programs should check that the data hashed into previous_leaf
+    /// matches the authority information necessary to execute this instruction.
+    pub fn replace_leaf(
+        ctx: Context<Modify>,
+        root: [u8; 32],
+        previous_leaf: [u8; 32],
+        new_leaf: [u8; 32],
+        index: u32,
+    ) -> Result<()> {
+        let mut merkle_roll_bytes = ctx.accounts.merkle_roll.try_borrow_mut_data()?;
+        let (header_bytes, rest) =
+            merkle_roll_bytes.split_at_mut(size_of::<ConcurrentMerkleTreeHeader>());
```

Comment:
The documentation say that this is checked in the instruction, but only the authority and signature is checked.  An owner check may be useful as well in case it's possible to no-op and trick someone into thinking that it worked

Reply:
Added